src/main/java/com/hivemq/mqtt/handler/connect/ConnectHandler.java

```diff
@@ -651,60 +651,58 @@ private void sendConnackSuccess(final @NotNull ChannelHandlerContext ctx, final
+        ThreadPreConditions.inNettyChildEventloop();
 
-            if (oldClient == null) {
-                channelPersistence.persist(msg.getClientIdentifier(), clientConnection);
-                return Futures.immediateFuture(null);
-            }
-            final ClientConnection oldClientConnection = oldClient.attr(ChannelAttributes.CLIENT_CONNECTION).get();
-            final SettableFuture<Void> disconnectFuture = oldClientConnection.getDisconnectFuture();
-            if (disconnectFuture == null) {
-                return Futures.immediateFailedFuture(new IllegalStateException("disconnect future must be present"));
-            }
-            // We have to check if the old client is currently taken over
-            // Otherwise we could take over the same client twice
-            final int nextRetry;
-            if (oldClientConnection.getClientState() != ClientState.DISCONNECTING
-                    && !oldClientConnection.getClientState().disconnected()) {
-
-                oldClientConnection.proposeClientState(ClientState.DISCONNECTING);
-
-                oldClientConnection.getChannel().eventLoop().execute(() ->
-                        mqttServerDisconnector.disconnect(oldClientConnection.getChannel(),
-                                "Disconnecting already connected client with id {} and ip {} because another client connects with that id",
-                                ReasonStrings.DISCONNECT_SESSION_TAKEN_OVER,
-                                Mqtt5DisconnectReasonCode.SESSION_TAKEN_OVER,
-                                ReasonStrings.DISCONNECT_SESSION_TAKEN_OVER));
-
-                nextRetry = retry;
-            } else {
-                // The client is currently taken over
-                if (retry >= MAX_TAKEOVER_RETRIES) {
-                    return Futures.immediateFailedFuture(new RuntimeException("maximum takeover retries exceeded"));
-                }
-                nextRetry = retry + 1;
-            }
-            final SettableFuture<Void> resultFuture = SettableFuture.create();
-            Futures.addCallback(disconnectFuture, new FutureCallback<>() {
-                @Override
-                public void onSuccess(final Void result) {
-                    resultFuture.setFuture(disconnectClientWithSameClientId(msg, clientConnection, nextRetry));
-                }
+        if (clientConnection.getClientState().disconnected()) {
+            return Futures.immediateFailedFuture(new RuntimeException("Disconnected before takeover."));
+        }
 
-                @Override
-                public void onFailure(final @NotNull Throwable t) {
-                    resultFuture.setException(t);
-                }
-            }, clientConnection.getChannel().eventLoop());
-            return resultFuture;
+        final ClientConnection persistedClientConnection = channelPersistence.tryPersist(msg.getClientIdentifier(), clientConnection);
```

Comment:
What do you think about calling this `persistIfAbsent` to be more explicit?

Reply:
And why do we not return a boolean here? 
So if we have successfully written our `clientConnection` to the persistence, it would return true and we don't have to do the next check.

Reply:
```java
if (channelPersistence.persistIfAbsent(msg.getClientIdentifier(), clientConnection)) {
    return Futures.immediateFuture(null);
}
```